3rdParty/iresearch/core/formats/columnstore2.cpp

```diff
@@ -801,82 +909,176 @@ doc_iterator::ptr fixed_length_column::iterator(bool /*consolidation*/) const {
+    auto& hdr = mutable_header();
+    if (!is_encrypted(hdr)) {
+      if (make_buffered_data<false>(hdr, in, blocks_, column_data_,
+                                    next_sorted_columns, nullptr)) {
+        buffered_input_ = std::make_unique<irs::bytes_ref_input>(
+            irs::bytes_ref{column_data_.data(), column_data_.size()});
+      }
+    } else {
+      irs::remapped_bytes_ref_input::mapping mapping;
+      if (make_buffered_data<true>(hdr, in, blocks_, column_data_,
+                                   next_sorted_columns, &mapping)) {
+        buffered_input_ = std::make_unique<irs::remapped_bytes_ref_input>(
+            irs::bytes_ref{column_data_.data(), column_data_.size()},
+            std::move(mapping));
+      }
+    }
+    if (buffered_input_) {
+      reset_stream(buffered_input_.get());
+    }
+  }
+
  private:
+  static std::vector<column_block> read_blocks_sparse(const column_header& hdr,
+                                                      index_input& in);
+
   template<typename ValueReader>
   class payload_reader : private ValueReader {
    public:
     template<typename... Args>
-    payload_reader(
-        const column_block* blocks,
-        Args&&... args)
-      : ValueReader{ std::forward<Args>(args)... },
-        blocks_{blocks} {
-    }
+    payload_reader(const column_block* blocks, Args&&... args)
+        : ValueReader{std::forward<Args>(args)...}, blocks_{blocks} {}
 
     bytes_ref payload(doc_id_t i);
 
    private:
     const column_block* blocks_;
-  }; // payload_reader
+  };  // payload_reader
+
+  template<bool encrypted>
+  bool make_buffered_data(
+      column_header& hdr, irs::index_input& in,
+      std::vector<column_block>& blocks,
+      std::vector<irs::byte_type>& column_data,
+      range<std::unique_ptr<column_reader>> next_sorted_columns,
+      remapped_bytes_ref_input::mapping* mapping) {
+    // idx adr/block offset length source
+    std::vector <std::tuple<size_t, bool, size_t, size_t, size_t>> chunks;
+    size_t chunks_size{0};
+    size_t block_idx{0};
+    std::vector<irs::byte_type> addr_buffer;
+    for (auto& block : blocks) {
+      if (bitpack::ALL_EQUAL == block.bits) {
+        size_t length = block.avg * block.last;
+        length += block.last_size;
+        chunks.emplace_back(block_idx++, false, chunks_size, length, block.data);
+        chunks_size += length;
+      } else {
+        // addr table size
+        const size_t addr_length = packed::bytes_required_64(
+            math::ceil64((block.last + 1), packed::BLOCK_SIZE_64), block.bits);
+        chunks.emplace_back(block_idx, true, chunks_size, addr_length, block.addr);
+        chunks_size += addr_length;
+        // block data size is calculated as end of the data for the last document in block
+        const size_t block_size = block.bits * sizeof(uint64_t);
+        auto const value_index = block.last % packed::BLOCK_SIZE_64;
+        addr_buffer.resize(block_size);
```

Comment:
The same `addr_buffer` object is reused here between multiple iterations.
It is not cleared between iterations. I guess this is not a problem because we are filling it starting at `addr_buffer.data()` in each iteration, but please double-check.

Reply:
Yes, here it is ok. I just want to avoid allocations for each iteration.